### PR TITLE
Introduce configuration file and initial setup through CLI

### PR DESCRIPTION
**Pull request by da-kami**
PRs chained on top of this one:

- [x] https://github.com/comit-network/xmr-btc-swap/pull/163 Configurable config with `--config` option on CLI (we will need this for swaps with ourselves...)
- [x] ~~https://github.com/comit-network/xmr-btc-swap/pull/165 Reset config with `--reset-config` option on CLI (this makes it easier for the user to clean up)~~
- [x] https://github.com/comit-network/xmr-btc-swap/pull/166 data-dir, in accordance to config-dir, should be pointed to the appropriate default directory using `directory-next`

Looks n' feel:

Initial startup:

```
2021-01-28T01:35:45.000205Z  INFO swap::trace: Initialized tracing with level: INFO
2021-01-28T01:35:45.001286Z  INFO swap: Database and Seed will be stored in directory: /Users/dakami/CoBloX/TEMP/swap-data-dir/alice
2021-01-28T01:35:45.003391Z  INFO swap::config::seed: Read in seed from file: /Users/dakami/CoBloX/TEMP/swap-data-dir/alice/seed.pem
2021-01-28T01:35:45.003603Z  INFO swap::config: Config file not found, running initial setup...

? Enter Bitcoind URL (including username and password if applicable) or hit return to use default (http://127.0.0.1:18332) › http://dakami:xkz4nyywpKf3BigwKIRdVijmWzaHCOmUisepQpDlsXnhpNd6uO@127.0.0.1:18332/
✔ Enter Bitcoind URL (including username and password if applicable) or hit return to use default · http://dakami:xkz4nyywpKf3BigwKIRdVijmWzaHCOmUisepQpDlsXnhpNd6uO@127.0.0.1:18332/
? Enter Bitcoind wallet name › alice
✔ Enter Bitcoind wallet name · alice
? Enter Monero Wallet RPC URL or hit enter to use default (http://127.0.0.1:38083/json_rpc) › 
✔ Enter Monero Wallet RPC URL or hit enter to use default · http://127.0.0.1:38083/json_rpc

2021-01-28T01:35:58.553401Z  INFO swap::config: Initial setup complete, config file created at /Users/dakami/Library/Application Support/xmr-btc-swap/config.toml 
2021-01-28T01:35:58.647761Z  INFO swap: Connection to Bitcoin wallet succeeded, balance: 0.00744521 BTC
2021-01-28T01:35:58.650060Z  INFO swap: Connection to Monero wallet succeeded, balance: 29.095359550000 XMR
2021-01-28T01:35:58.650258Z  INFO swap: Swap sending 0.030000000000 XMR and receiving 0.00060000 BTC started with ID e07b2cc1-3749-48fd-931a-6cbaf57b8124
2021-01-28T01:35:59.004306Z  INFO swap::protocol::alice::swap: Current state:started
```

After:
```
2021-01-28T01:36:57.881654Z  INFO swap::trace: Initialized tracing with level: INFO
2021-01-28T01:36:57.882691Z  INFO swap: Database and Seed will be stored in directory: /Users/dakami/CoBloX/TEMP/swap-data-dir/alice
2021-01-28T01:36:57.884171Z  INFO swap::config::seed: Read in seed from file: /Users/dakami/CoBloX/TEMP/swap-data-dir/alice/seed.pem
2021-01-28T01:36:57.884353Z  INFO swap::config: Using config file at default path: /Users/dakami/Library/Application Support/xmr-btc-swap/config.toml
2021-01-28T01:36:57.996153Z  INFO swap: Connection to Bitcoin wallet succeeded, balance: 0.00744521 BTC
2021-01-28T01:36:57.998648Z  INFO swap: Connection to Monero wallet succeeded, balance: 29.095359550000 XMR
2021-01-28T01:36:57.998928Z  INFO swap: Swap sending 0.030000000000 XMR and receiving 0.00060000 BTC started with ID 08dd6dc1-9460-4c0a-91ef-a05df309b6ed
2021-01-28T01:36:58.353738Z  INFO swap::protocol::alice::swap: Current state:started
```

Command:

```
run --package swap --bin swap -- 
  --data-dir /Users/dakami/CoBloX/TEMP/swap-data-dir/alice 
  sell-xmr 
    --receive-btc 0.0006 
    --send-xmr 0.03
```

### CODE REVIEW

File: swap/src/config.rs [hunk 1]
```diff
@@ -1,110 +1,146 @@
+use crate::fs::ensure_directory_exists;
```
Comment (D4nte) [inline swap/src/config.rs:1]: This diff is unreadable. If you strongly feel that the swap CLI parameters to access the node must be named `config` and the blockchain characteristics and protocol parameters must be named `settings` I would recommend the following approach:

- introduce new module `configuration` in this PR for the node urls.
- Possible rename the `config` module to `settings`
- Do a follow-up PR where you rename `configuration` to `config`.

The criss-cross rename should be avoided for the sake of the reviewers.

Regarding the terminology for the blockchain times and expiries, the terminology is tricky and I believe it is because we mixed up some concepts. We are currently storing few types of information:
1. Blockchain characteristics/behaviour: avg block time, network
2. Security parameters: number of confirmation to finality, bob time to act
3. Protocol parameters: Bitcoin expiries

I think a bigger revamp to be done to separate those 3 concepts.

If you want a generic name I'd go for `parameters`.

File: swap/src/fs.rs [hunk 2]
```diff
@@ -1,4 +1,21 @@
-use std::path::Path;
+use anyhow::Context;
+use directories_next::ProjectDirs;
+use std::path::{Path, PathBuf};
+
+/// This is to store the configuration and seed files
+// Linux: /home/<user>/.config/xmr-btc-swap/
+// OSX: /Users/<user>/Library/Preferences/xmr-btc-swap/
+#[allow(dead_code)]
```
Comment (D4nte) [inline swap/src/fs.rs:8]: Why is this needed?

File: swap/src/main.rs [hunk 3]
```diff
@@ -12,12 +12,18 @@
...
 #![allow(non_snake_case)]
 
-use crate::cli::{Command, Options, Resume};
+use crate::{
+    cli::{Command, Options, Resume},
+    config::{
+        initial_setup, query_user_for_initial_testnet_config, read_config, ConfigNotInitialized,
+    },
+};
 use anyhow::{Context, Result};
```
Comment (D4nte) [inline swap/src/main.rs:21]:
```suggestion
use cli::{Command, Options, Resume};
use config::{
        initial_setup, query_user_for_initial_testnet_config, read_config, ConfigNotInitialized,
    };
use anyhow::{Context, Result};
```
The usage of `use crate::` in the `main.rs` is strange (I know it was here before).

File: swap/src/settings.rs [hunk 4]
```diff
@@ -0,0 +1,177 @@
+use crate::{bitcoin::Timelock, config::File};
+use conquer_once::Lazy;
+use std::time::Duration;
+use url::Url;
+
+pub struct Settings {
+    pub wallets: Wallets,
+    pub protocol: Protocol,
+}
```
Comment (D4nte) [inline swap/src/settings.rs:9]: I don't think that the wallet configuration and protocol settings should be bundled together.

The wallets settings are only used to instantiate the wallets where the protocol ones are carried around.

**Review by D4nte — COMMENTED** [review]
Can you please tackle the naming comments so it is easier to review?

File: swap/src/config.rs (hunk 1 above)
Comment (da-kami) [reply swap/src/config.rs:1]: > This diff is unreadable. 

I split the changes into separate, isolated commits. Yes, the complete diff turned out to be ugly, but could you by any change review by commits?

File: swap/src/fs.rs (hunk 2 above)
Comment (da-kami) [reply swap/src/fs.rs:8]: Because of lib and main both referencing fs (which is unnecessary). I fixed this here: https://github.com/comit-network/xmr-btc-swap/pull/166

I think putting all the PRs up was probably not a good idea, it's getting cumbersome to align things.

File: swap/src/main.rs (hunk 3 above)
Comment (da-kami) [reply swap/src/main.rs:21]: This is because of the lib as well. `config` is ambiguous, because there is we have a dependency named `config` as well, and `swap::config` causes compile issues.

File: swap/src/config.rs (hunk 1 above)
Comment (da-kami) [reply swap/src/config.rs:1]: > If you strongly feel that the swap CLI parameters to access the node must be named config and the blockchain characteristics and protocol parameters must be named settings I would recommend the following approach.

I don't feel strong about it per-se, but it follows the idea of our other projects and is based on https://github.com/mehcode/config-rs as far as I could see.

File: swap/src/settings.rs (hunk 4 above)
Comment (da-kami) [reply swap/src/settings.rs:9]: I did that to ensure that both are initialized for `testnet`. If we separate it, we could run into the issue of mixing networks. But that is probably not enough to justify the `Settings` struct. I can remove it once we merged the changes on top into this PR.

File: swap/src/config.rs (hunk 1 above)
Comment (da-kami) [reply swap/src/config.rs]: I tried to change the `settings` and `config` from the bottom up but it is quite a pain. This accounts to https://github.com/comit-network/xmr-btc-swap/pull/161#discussion_r566509040 as well. 

I think the current design is fine, and the commits are contained at the moment. Please review by commit if possible @D4nte - all commits should be contained now. 

As for the split between settings and config: I do think it is a good split because it emphasizes config (as in config-file) vs what is to be used within the program.
As for separating the wallet settings from the execution parameters: Those are related, because the wallets are needed for execution as well. Eventually we might move execution settings into the config as well. I also think that creating overall `testnet` settings is not wrong.

Would you be OK with merging this as is @D4nte or do you insist on changes?

File: swap/src/bitcoin/wallet.rs [hunk 5]
```diff
@@ -170,17 +170,21 @@ impl TransactionBlockHeight for Wallet {
...
+        &self,
+        txid: Txid,
+        settings: settings::Protocol,
+    ) -> Result<()> {
         // TODO(Franck): This assumes that bitcoind runs with txindex=1
 
         // Divide by 4 to not check too often yet still be aware of the new block early
         // on.
-        let mut interval = interval(config.bitcoin_avg_block_time / 4);
+        let mut interval = interval(settings.bitcoin_avg_block_time / 4);
```
Comment (D4nte) [inline swap/src/bitcoin/wallet.rs:182]: This is strange. The bitcoin avg block time is not really a "setting"

File: swap/src/config.rs [hunk 6]
```diff
@@ -1,110 +1 @@
 pub mod seed;
```
Comment (D4nte) [inline swap/src/config.rs:14]: Did you do a `git mv`? shame github does not catch it if you did.

File: swap/src/config.rs [hunk 7]
```diff
@@ -1 +1,149 @@
...
+use url::Url;
+
 pub mod seed;
+
+const DEFAULT_BITCOIND_TESTNET_URL: &str = "http://127.0.0.1:18332";
+const DEFAULT_MONERO_WALLET_RPC_TESTNET_URL: &str = "http://127.0.0.1:38083/json_rpc";
+
+#[derive(Clone, Debug, serde::Serialize, serde::Deserialize, PartialEq)]
+#[serde(deny_unknown_fields)]
+pub struct File {
```
Comment (D4nte) [inline swap/src/config.rs:20]: If this module is about defining the `File` struct then it should be called `file.rs`.

If you compare to the other repos, we have
- `config/file.rs`: config file serialization/deserialization
- `config/settings.rs`: the settings to pass around
- `config/mod.rs`: types used in both `file.rs` and `settings.rs`.

File: swap/src/fs.rs (hunk 2 above)
Comment (D4nte) [reply swap/src/fs.rs]: #166 is merged, why is the dead code attribute still here?

File: swap/src/config.rs [hunk 8]
```diff
@@ -49,27 +49,24 @@ pub struct Monero {
...
-#[derive(thiserror::Error, Debug)]
-pub enum Error {
-    #[error("config not initialized")]
-    ConfigNotInitialized,
-    #[error("other error")]
-    Other(#[from] anyhow::Error),
-}
+#[derive(thiserror::Error, Debug, Clone, Copy)]
+#[error("config not initialized")]
+pub struct ConfigNotInitialized {}
```
Comment (D4nte) [inline swap/src/config.rs:53]: Do you mean that the configuration is missing?

File: swap/src/main.rs [hunk 9]
```diff
@@ -206,13 +207,12 @@ async fn main() -> Result<()> {
...
 fn init_settings() -> Result<Settings> {
     let config_path = default_config_path()?;
-    let config = match read_config(config_path.clone()) {
+    let config = match read_config(config_path.clone())? {
         Ok(config) => config,
-        Err(config::Error::ConfigNotInitialized) => {
+        Err(ConfigNotInitialized {}) => {
             initial_setup(config_path.clone(), query_user_for_initial_testnet_config)?;
-            read_config(config_path)?
+            read_config(config_path)?.expect("after initial setup config can be read")
```
Comment (D4nte) [inline swap/src/main.rs:229]: `init_settings` returns a result. Return a result instead of using `expect`.

File: swap/src/fs.rs [hunk 10]
```diff
@@ -5,12 +5,10 @@ use std::path::{Path, PathBuf};
 /// This is to store the configuration and seed files
 // Linux: /home/<user>/.config/xmr-btc-swap/
 // OSX: /Users/<user>/Library/Preferences/xmr-btc-swap/
-#[allow(dead_code)]
```
Comment (D4nte) [inline swap/src/fs.rs]: This line is added and then deleted.

**Review by D4nte — COMMENTED** [review]
The module structure is not on point.
the `config` module contains the serializable/deserializable `File` struct.
the `settings` module contains the struct that holds the user settings, the protocol parameters and the blockchain characteristics.

also, `settings` and `config` are in totally separate modules.

File: swap/src/config.rs (hunk 8 above)
Comment (da-kami) [reply swap/src/config.rs:53]: It is not initialized. The current process is: Upon initial startup the user is queried for the RPC urls and wallet name (bitcoin only as long as we are using bitcoind's wallet) and the config is written automatically. 

We can rename it to `ConfigMissing`, but it is only used for triggering the `initial_setup` at the moment, so I think the name `ConfigNotInitialized` expresses exactly what it is.

File: swap/src/main.rs (hunk 9 above)
Comment (da-kami) [reply swap/src/main.rs:233]: It returns a nested result, where the outer is an actual error we want to bubble up, and the inner is `ConfigNotInitialized`. After running `initial_setup` we expect the config to be present. The expect makes this explicit.

File: swap/src/config.rs (hunk 7 above)
Comment (da-kami) [reply swap/src/config.rs:20]: I combined all relevant config specific code in the module `config.rs` (which serves as module because we use module name over folders and `mod.rs`). The `config.rs` module does not only contain the `File` struct, but also functions to read and initialize the config. I can separate `File` into a separate `file` module under `config`, but I did not see the necessity yet.

**Review by D4nte — APPROVED** [review]
